### ISSUE TITLE
 Issue with Rank=0: ResNet56 Training on cifar10

### ISSUE THREAD

**Issue by Xuzhuobin** [MaxBourdon/mars#2]
Thank you for sharing your work. I encountered an issue while replicating the experiments of ResNet on cifar10 from your paper. The specific situation is as follows:

When training the mask -TT ResNet56 on the cifar10 dataset with parameters (init_rank=20, pi=1e-2, alpha=2.25 as per the paper setting), I observed instances where certain layers experienced rank=0 after applying the mask for like 90 epochs. This results in errors during forward inference. Additionally, even with the inclusion of a warm-up step in the initial training phase, the test accuracy of the model remains nearly unchanged for the many epochs.

I would like to inquire if the aforementioned issues could be attributed to suboptimal initialization or whether there are additional details that need to be added during the training process to prevent the occurrence of the rank=0 situation. I am eagerly looking forward to and appreciative of your response.

**Comment by MaxBourdon** [comment]
Hi!
As far as I remember, these steps can help to train ResNet with MARS without the zero masks issue:
1. Setting logits_init >= 1.0, pi >= 0.1 for MARS layers.
2. Using Adam optimizer.
3. Warming up for ~10 epochs with a smaller LR ~1e-3.
4. Applying temperature annealing after each epoch as max(1e-2, 0.99 * t)
5. Sometimes using smaller or even no WD on MARS layers is also helpful but I don't think it's so necessary.
Generally, that should work with standard training using LR schedule (starting at ~1e-2) and WD (~1e-4). Also, remember that not all blocks in the network are "MARSed". Finally, don't hesitate to experiment, I am sure there are better protocols to ensure the best training and performance.
Please, let me know if you have any further troubles with that!

**Comment by Xuzhuobin** [comment]
Thanks,that helps me a lot.

**Comment by Xuzhuobin** [comment]
Thanks,that helps me a lot.
